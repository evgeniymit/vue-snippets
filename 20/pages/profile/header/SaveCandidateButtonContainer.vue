<template>
  <div class="candidate-header-secondary-flex-item">
    <div v-if="isDisabled" id="saveCandidateBtn" class="flex-bar-btn-item save">
      <responsive-tooltip :label="disabledTooltip">
        <a class="button is-default-icon is-default-icon-row is-primary" disabled="disabled">
          <div class="is-default-icon-item icon-btn-item save-icon">
            <svg-icon name="star-on" class="saved-icon"></svg-icon>
          </div>
          <div v-translate class="is-default-icon-item btn-title">Saved</div>
        </a>
      </responsive-tooltip>
    </div>
    <save-candidate-popover-container
      v-else
      :candidate="candidate"
      position="bottom"
      source="profilePage"
      @post-save="postSave()">

      <div id="saveCandidateBtn" class="flex-bar-btn-item save">
        <responsive-tooltip :label="saveThisCandidateLabel" :active="!saved">
          <a class="button is-default-icon is-default-icon-row">
            <div v-if="loading" class="loader save-loader"></div>
            <div v-else class="is-default-icon-item icon-btn-item save-icon">
              <svg-icon :name="saved ? 'star-on': 'star-off'" :class="saved ? 'saved-icon': 'not-saved-icon'"></svg-icon>
            </div>
            <div class="is-default-icon-item btn-title">{{ savedButtonText }}</div>
          </a>
        </responsive-tooltip>
      </div>

    </save-candidate-popover-container>
  </div>
</template>

<script>

import 'top20common/distribution/icons/star-on';
import 'top20common/distribution/icons/star-off';

import {mapGetters} from 'vuex';

import ResponsiveTooltip from '@/common/components/ResponsiveTooltip';
import SaveCandidatePopoverContainer from '@/common/save-candidate/SaveCandidatePopoverContainer';

export default {
  props: {
    candidate: Object
  },
  computed: {
    ...mapGetters([
      'unavailableSavedState',
      'contactedToAccountActiveProjectsByCandidateIdState',
      'userDetailsState',
      'accountUsersNameByIdState'
    ]),
    saved() {
      return this.$store.getters.savedToAccountActiveProjectsByCandidateIdState[this.candidate.id];
    },
    isDisabled() {
      return (
        this.contactedToAccountActiveProjectsByCandidateIdState[this.candidate.id] ||
        this.unavailableSavedState[this.candidate.id]
      );
    },
    loading() {
      return this.$store.getters.eachSavedLoadingState[this.candidate.id];
    },
    saveThisCandidateLabel() {
      return this.$gettext('Save this candidate');
    },
    disabledTooltip() {
      const candidate =
        this.unavailableSavedState[this.candidate.id] ||
        this.contactedToAccountActiveProjectsByCandidateIdState[this.candidate.id];
      if (candidate.author_id === this.userDetailsState.pk) {
        return this.$gettext("You've already added this candidate to a project");
      }
      return `${this.$gettext('Already added to a project by')} ${this.accountUsersNameByIdState[candidate.author_id]}`;
    },
    savedButtonText() {
      return this.saved ? this.$gettext('Saved') : this.$gettext('Save');
    }
  },
  methods: {
    postSave() {
      this.$emit('post-save');
    }
  },
  components: {SaveCandidatePopoverContainer, ResponsiveTooltip}
};
</script>

<style lang="scss" scoped>
@import '~VariablesStyles';

/deep/ .trigger {
  width: 100%;
}

/deep/ .v-popover {
  width: 100%;
}

.candidate-header-secondary-flex-item {
  display: flex;
  align-items: flex-start;
  .save-candidate-tooltip {
    display: flex;
    align-items: flex-start;
  }
  /deep/ .v-popover {
    display: flex;
    align-items: flex-start;
  }
}

.icon-btn-item {
  display: flex;
  align-items: center;
}

.not-saved-icon.svg-fill {
  color: $icons;
}
.saved-icon.svg-fill {
  color: $secondary;
}

.button.is-default-icon[disabled] {
  color: $white;
  cursor: unset;
  .saved-icon {
    color: $white;
  }
}

.button.is-default-icon:hover {
  .saved-icon {
    color: $white;
  }
  .not-saved-icon /deep/ path {
    stroke: $white;
  }
}

.save .is-default-icon.is-default-icon-row {
  width: 100px;
}

.button.is-default {
  background: $white;
  border: none;
  box-shadow: 0 0 10px 0 $white-light-opacity;
  color: $primary;
  float: right;
}

.flex-bar-btn-item {
  display: flex;
  align-items: center;
}
@media (min-width: $tablet) {
  .flex-bar-btn-item {
    margin-left: 25px;
  }
}
</style>
