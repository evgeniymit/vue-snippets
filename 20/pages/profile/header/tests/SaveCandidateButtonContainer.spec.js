import localVue from '@/common/test-helpers/mock-vue';
import { mount } from '@vue/test-utils'
import SaveCandidateButtonContainer from '@/pages/profile/header/SaveCandidateButtonContainer'
import Vuex from "vuex";


describe('SaveCandidateButtonContainer', () => {

  const oldMath = Object.create(global.Math);
  const mockMath = Object.create(global.Math);
  mockMath.random = () => 1;

  beforeAll(() => {
    global.Math = mockMath;
  });
  afterAll(() => {
    global.Math = oldMath;
  });

  describe('candidate is not saved or contacted by anyone', () => {
    it('the save button should be enabled', () => {
      const candidate = {id: 'PGT-888', categories: [] };
      const propsData = {candidate};
      const stubs = ['new-project-popup-container', 'modal'];
      const store = new Vuex.Store({
        getters: {
          unavailableSavedState() {return {}},
          savedToAccountActiveProjectsByCandidateIdState() {return {}},
          contactedToAccountActiveProjectsByCandidateIdState() {return {}},
          eachSavedLoadingState() {return {}},
          userValidIndustriesState() {return {}},
          userValidCitiesState() {return {}},
          userDetailsState() {return {pk: 'user_pk'}},
          accountUsersNameByIdState() {return {}},
          projectFormState() {return {}},
          allFiltersState() {return {}},
        }
      });
      const $archive = {getUnarchivedProjects: jest.fn()};
      const $filters = {getItemCollectionSeparatedByGroup: jest.fn()};
      const mocks = {$archive, $filters};
      const wrapper = mount(SaveCandidateButtonContainer, {localVue, propsData, stubs, store, mocks});

      expect(wrapper.vm.savedButtonText).toBe('Save');
      expect(wrapper.find('span.is-primary.is-top.is-medium').attributes()['data-label']).toBe('Save this candidate');
      const button = wrapper.find('a.button');
      expect(button.isVisible()).toBe(true);
      expect(button.attributes().disabled).toBeFalsy();
      expect(wrapper.html()).toMatchSnapshot();
    });
  });

  describe('candidate is saved by current user', () => {
    it('the save button should be enabled', () => {
      const candidate = {id: 'PGT-888', categories: [] };
      const allSaved = {'PGT-888': {candidate}};
      const propsData = {candidate};
      const stubs = ['new-project-popup-container', 'modal'];
      const store = new Vuex.Store({
        getters: {
          unavailableSavedState() {return {}},
          savedToAccountActiveProjectsByCandidateIdState() {return allSaved},
          contactedToAccountActiveProjectsByCandidateIdState() {return {}},
          eachSavedLoadingState() {return {}},
          userValidIndustriesState() {return {}},
          userValidCitiesState() {return {}},
          userDetailsState() {return {pk: 'user_pk'}},
          accountUsersNameByIdState() {return {'user_pk': 'John Doe'}},
          projectFormState() {return {}},
          allFiltersState() {return {}},
        }
      });
      const $archive = {getUnarchivedProjects: jest.fn()};
      const $filters = {getItemCollectionSeparatedByGroup: jest.fn()};
      const mocks = {$archive, $filters};
      const wrapper = mount(SaveCandidateButtonContainer, {localVue, propsData, stubs, store, mocks});

      expect(wrapper.vm.savedButtonText).toBe('Saved');
      expect(wrapper.find('span.is-primary.is-top.is-medium').attributes()['data-label']).toBe('Save this candidate');
      const button = wrapper.find('a.button');
      expect(button.isVisible()).toBe(true);
      expect(button.attributes().disabled).toBeFalsy();
      expect(wrapper.html()).toMatchSnapshot();
    });
  });

  describe('candidate is contacted by current user', () => {
    it('the save button should be disabled', () => {
      const candidate = {id: 'PGT-888', categories: [] };
      const allContacted = {'PGT-888': {candidate, author_id: 'user_pk'}};
      const propsData = {candidate};
      const stubs = ['new-project-popup-container', 'modal'];
      const store = new Vuex.Store({
        getters: {
          unavailableSavedState() {return {}},
          savedToAccountActiveProjectsByCandidateIdState() {return {}},
          contactedToAccountActiveProjectsByCandidateIdState() {return allContacted},
          eachSavedLoadingState() {return {}},
          userValidIndustriesState() {return {}},
          userValidCitiesState() {return {}},
          userDetailsState() {return {pk: 'user_pk'}},
          accountUsersNameByIdState() {return {'user_pk': 'John Doe'}},
          projectFormState() {return {}},
          allFiltersState() {return {}},
        }
      });
      const $archive = {getUnarchivedProjects: jest.fn()};
      const $filters = {getItemCollectionSeparatedByGroup: jest.fn()};
      const mocks = {$archive, $filters};
      const wrapper = mount(SaveCandidateButtonContainer, {localVue, propsData, stubs, store, mocks});

      expect(wrapper.vm.savedButtonText).toBe('Save');
      expect(wrapper.find('span.is-primary.is-top.is-medium').attributes()['data-label']).toBe('You\'ve already added this candidate to a project');
      const button = wrapper.find('a.button');
      expect(button.isVisible()).toBe(true);
      expect(button.attributes().disabled).toBeTruthy();
      expect(wrapper.html()).toMatchSnapshot();
    });
  });

  describe('candidate is in unavailableSavedState (saved by other user)', () => {
    it('the save button should be disabled', () => {
      const candidate = {id: 'PGT-888', categories: [] };
      const unavailableSavedState = {'PGT-888': {candidate, author_id: 'other_user_pk'}};
      const propsData = {candidate};
      const stubs = ['new-project-popup-container', 'modal'];
      const store = new Vuex.Store({
        getters: {
          unavailableSavedState() {return unavailableSavedState},
          savedToAccountActiveProjectsByCandidateIdState() {return {}},
          contactedToAccountActiveProjectsByCandidateIdState() {return {}},
          eachSavedLoadingState() {return {}},
          userValidIndustriesState() {return {}},
          userValidCitiesState() {return {}},
          userDetailsState() {return {pk: 'user_pk'}},
          accountUsersNameByIdState() {return {'user_pk': 'John Doe', 'other_user_pk': 'Maymay'}},
          projectFormState() {return {}},
          allFiltersState() {return {}},
        }
      });
      const $archive = {getUnarchivedProjects: jest.fn()};
      const $filters = {getItemCollectionSeparatedByGroup: jest.fn()};
      const mocks = {$archive, $filters};
      const wrapper = mount(SaveCandidateButtonContainer, {localVue, propsData, stubs, store, mocks});

      expect(wrapper.vm.savedButtonText).toBe('Save');
      expect(wrapper.find('span.is-primary.is-top.is-medium').attributes()['data-label']).toBe('Already added to a project by Maymay');
      const button = wrapper.find('a.button');
      expect(button.isVisible()).toBe(true);
      expect(button.attributes().disabled).toBeTruthy();
      expect(wrapper.html()).toMatchSnapshot();
    });
  });
  describe('candidate is in contactedToAccountActiveProjectsByCandidateIdState (contacted by any user)', () => {
    it('the save button should be disabled', () => {
      const candidate = {id: 'PGT-888', categories: [] };
      const allContacted = {'PGT-888': {candidate, author_id: 'other_user_pk'}};
      const propsData = {candidate};
      const stubs = ['new-project-popup-container', 'modal'];
      const store = new Vuex.Store({
        getters: {
          unavailableSavedState() {return {}},
          savedToAccountActiveProjectsByCandidateIdState() {return {}},
          contactedToAccountActiveProjectsByCandidateIdState() {return allContacted},
          eachSavedLoadingState() {return {}},
          userValidIndustriesState() {return {}},
          userValidCitiesState() {return {}},
          userDetailsState() {return {pk: 'user_pk'}},
          accountUsersNameByIdState() {return {'user_pk': 'John Doe', 'other_user_pk': 'Maymay'}},
          projectFormState() {return {}},
          allFiltersState() {return {}},
        }
      });
      const $archive = {getUnarchivedProjects: jest.fn()};
      const $filters = {getItemCollectionSeparatedByGroup: jest.fn()};
      const mocks = {$archive, $filters};
      const wrapper = mount(SaveCandidateButtonContainer, {localVue, propsData, stubs, store, mocks});

      expect(wrapper.vm.savedButtonText).toBe('Save');
      expect(wrapper.find('span.is-primary.is-top.is-medium').attributes()['data-label']).toBe('Already added to a project by Maymay');
      const button = wrapper.find('a.button');
      expect(button.isVisible()).toBe(true);
      expect(button.attributes().disabled).toBeTruthy();
      expect(wrapper.html()).toMatchSnapshot();
    });
  });
});
