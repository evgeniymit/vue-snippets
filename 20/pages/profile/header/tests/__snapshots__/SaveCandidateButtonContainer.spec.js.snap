// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SaveCandidateButtonContainer candidate is contacted by current user the save button should be disabled 1`] = `
<div class="candidate-header-secondary-flex-item">
  <div id="saveCandidateBtn" class="flex-bar-btn-item save"><span data-label="You've already added this candidate to a project" class="is-primary is-top is-medium tooltip"><a disabled="disabled" class="button is-default-icon is-default-icon-row is-primary"><div class="is-default-icon-item icon-btn-item save-icon"><svg data-v-7ef413f4="" version="1.1" viewBox="0 0 16 16" class="saved-icon svg-icon svg-fill"><g xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink"><path fill-rule="evenodd" d="M8 13.267L3.056 16 4 10.211l-4-4.1 5.528-.844L8 0l2.472 5.267L16 6.11l-4 4.1.944 5.789z"></path></g></svg></div> <div class="is-default-icon-item btn-title" data-msgid="Saved" data-current-language="en_US">Saved</div></a></span></div>
</div>
`;

exports[`SaveCandidateButtonContainer candidate is in contactedToAccountActiveProjectsByCandidateIdState (contacted by any user) the save button should be disabled 1`] = `
<div class="candidate-header-secondary-flex-item">
  <div id="saveCandidateBtn" class="flex-bar-btn-item save"><span data-label="Already added to a project by Maymay" class="is-primary is-top is-medium tooltip"><a disabled="disabled" class="button is-default-icon is-default-icon-row is-primary"><div class="is-default-icon-item icon-btn-item save-icon"><svg data-v-7ef413f4="" version="1.1" viewBox="0 0 16 16" class="saved-icon svg-icon svg-fill"><g xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink"><path fill-rule="evenodd" d="M8 13.267L3.056 16 4 10.211l-4-4.1 5.528-.844L8 0l2.472 5.267L16 6.11l-4 4.1.944 5.789z"></path></g></svg></div> <div class="is-default-icon-item btn-title" data-msgid="Saved" data-current-language="en_US">Saved</div></a></span></div>
</div>
`;

exports[`SaveCandidateButtonContainer candidate is in unavailableSavedState (saved by other user) the save button should be disabled 1`] = `
<div class="candidate-header-secondary-flex-item">
  <div id="saveCandidateBtn" class="flex-bar-btn-item save"><span data-label="Already added to a project by Maymay" class="is-primary is-top is-medium tooltip"><a disabled="disabled" class="button is-default-icon is-default-icon-row is-primary"><div class="is-default-icon-item icon-btn-item save-icon"><svg data-v-7ef413f4="" version="1.1" viewBox="0 0 16 16" class="saved-icon svg-icon svg-fill"><g xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink"><path fill-rule="evenodd" d="M8 13.267L3.056 16 4 10.211l-4-4.1 5.528-.844L8 0l2.472 5.267L16 6.11l-4 4.1.944 5.789z"></path></g></svg></div> <div class="is-default-icon-item btn-title" data-msgid="Saved" data-current-language="en_US">Saved</div></a></span></div>
</div>
`;

exports[`SaveCandidateButtonContainer candidate is not saved or contacted by anyone the save button should be enabled 1`] = `
<div class="candidate-header-secondary-flex-item">
  <div class="v-popover"><span aria-describedby="popover_" tabindex="-1" class="trigger" style="display: inline-block;"><div><div id="saveCandidateBtn" class="flex-bar-btn-item save"><span data-label="Save this candidate" class="is-primary is-top is-medium tooltip"><a class="button is-default-icon is-default-icon-row"><div class="is-default-icon-item icon-btn-item save-icon"><svg data-v-7ef413f4="" version="1.1" viewBox="0 0 16 16" class="svg-icon svg-fill not-saved-icon"><g xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink"><path fill="none" fill-rule="evenodd" stroke="#AAA" d="M12.285 15.064l-.82-5.021 3.498-3.584-4.83-.738L8 1.177 5.867 5.72l-4.83.738 3.497 3.584-.819 5.021L8 12.696l4.285 2.368z"></path></g></svg></div> <div class="is-default-icon-item btn-title">Save</div></a></span></div>
</div> </span> </div>
</div>
`;

exports[`SaveCandidateButtonContainer candidate is saved by current user the save button should be enabled 1`] = `
<div class="candidate-header-secondary-flex-item">
  <div class="v-popover"><span aria-describedby="popover_" tabindex="-1" class="trigger" style="display: inline-block;"><div><div id="saveCandidateBtn" class="flex-bar-btn-item save"><span data-label="Save this candidate" class="is-primary is-top is-medium"><a class="button is-default-icon is-default-icon-row"><div class="is-default-icon-item icon-btn-item save-icon"><svg data-v-7ef413f4="" version="1.1" viewBox="0 0 16 16" class="svg-icon svg-fill saved-icon"><g xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink"><path fill-rule="evenodd" d="M8 13.267L3.056 16 4 10.211l-4-4.1 5.528-.844L8 0l2.472 5.267L16 6.11l-4 4.1.944 5.789z"></path></g></svg></div> <div class="is-default-icon-item btn-title">Saved</div></a></span></div>
</div> </span> </div>
</div>
`;
